sycl/source/detail/graph_impl.hpp

```diff
@@ -1236,7 +1172,12 @@ class graph_impl : public std::enable_shared_from_this<graph_impl> {
   /// @param Node The node to add deps for
   /// @param Deps List of dependent nodes
   void addDepsToNode(std::shared_ptr<node_impl> Node,
-                     const std::vector<std::shared_ptr<node_impl>> &Deps) {
+                     std::vector<std::shared_ptr<node_impl>> &Deps) {
+    // Remove empty shared pointers from the list
+    auto EmptyElementIter =
+        std::remove(Deps.begin(), Deps.end(), std::shared_ptr<node_impl>());
+    Deps.erase(EmptyElementIter, Deps.end());
```

Comment:
Curious how empty shared pointers end up in the dependency list, is that a problem we can fix?

Reply:
It's this `sycl/source/handler.cpp` part of the PR diff
```diff
-      NodeImpl = DependentNode
-                     ? GraphImpl->add(NodeType, std::move(CommandGroup),
-                                      {DependentNode})
-                     : GraphImpl->add(NodeType, std::move(CommandGroup));
+      std::vector<std::shared_ptr<ext::oneapi::experimental::detail::node_impl>>
+          Deps = {DependentNode};
+      NodeImpl = GraphImpl->add(NodeType, std::move(CommandGroup), Deps);
```

If the queue is in order we, get `auto DependentNode = GraphImpl->getLastInorderNode(MQueue);` to use as a dependency, which can return `nullptr`.

To avoid copying the `Deps` containter (see Fábio's feedback at https://github.com/reble/llvm/pull/380#discussion_r1805244997) I changed the `add()` API to take a reference to a vector. A nullptr can therefor appear in the vector.

Things we could do to avoid this are:
* use a const cast as per Fábio's original feedback.
* use `std::optional` in some way

I wasn't convinced that either would be better code, but happy to use them (or other ideas) if there is a preference.